File: server/src/modules/system/organization/organization.controller.ts
import { Context } from "koa";
import { Result } from "@/core/result";
import organizationService from "./organization.service";
import { AddManagersDto, AddMembersDto } from "./organization.dto";

function currentUserId(ctx: Context): number {
  return (ctx.state as any).user?.id as number;
}
function isSuperAdmin(ctx: Context): boolean {
  const result = (ctx.state as any).user?.roles ?? [];
  return result.includes("R_SUPER") || result.includes("R_ADMIN");
}

class OrganizationController {
  async getTree(ctx: Context) {
    ctx.body = Result.success(await organizationService.getTree());
  }

  async getAllUsers(ctx: Context) {
    ctx.body = Result.success(await organizationService.getAllUsers());
  }

  async getManagers(ctx: Context) {
    const deptId = Number(ctx.params.id);
    ctx.body = Result.success(await organizationService.getManagers(deptId));
  }

  async addManagers(ctx: Context) {
    if (!isSuperAdmin(ctx)) {
      ctx.status = 403;
      ctx.body = Result.error("无权限：仅超级管理员可添加管理者", 403);
      return;
    }
    const deptId = Number(ctx.params.id);
    const parsed = AddManagersDto.safeParse(ctx.request.body);
    if (!parsed.success) {
      ctx.body = Result.error(parsed.error.errors[0]?.message ?? "参数错误");
      return;
    }
    await organizationService.addManagers(deptId, parsed.data.userIds);
    ctx.body = Result.success(null, "添加管理者成功");
  }

  async removeManager(ctx: Context) {
    if (!isSuperAdmin(ctx)) {
      ctx.status = 403;
      ctx.body = Result.error("无权限：仅超级管理员可移除管理者", 403);
      return;
    }
    const deptId = Number(ctx.params.id);
    const userId = Number(ctx.params.userId);
    await organizationService.removeManager(deptId, userId);
    ctx.body = Result.success(null, "已移除管理者");
  }

  async getMembers(ctx: Context) {
    const deptId = Number(ctx.params.id);
    ctx.body = Result.success(await organizationService.getMembers(deptId));
  }

  async addMembers(ctx: Context) {
    const deptId = Number(ctx.params.id);
    const uid = currentUserId(ctx);
    const ok =
      isSuperAdmin(ctx) || (await organizationService.isManager(deptId, uid));
    if (!ok) {
      ctx.status = 403;
      ctx.body = Result.error("无权限：您不是该部门的管理者", 403);
      return;
    }
    const parsed = AddMembersDto.safeParse(ctx.request.body);
    if (!parsed.success) {
      ctx.body = Result.error(parsed.error.errors[0]?.message ?? "参数错误");
      return;
    }
    await organizationService.addMembers(deptId, parsed.data.userIds);
    ctx.body = Result.success(null, "添加成员成功");
  }

  async removeMember(ctx: Context) {
    const deptId = Number(ctx.params.id);
    const uid = currentUserId(ctx);
    const ok =
      isSuperAdmin(ctx) || (await organizationService.isManager(deptId, uid));
    if (!ok) {
      ctx.status = 403;
      ctx.body = Result.error("无权限：您不是该部门的管理者", 403);
      return;
    }
    const userId = Number(ctx.params.userId);
    await organizationService.removeMember(deptId, userId);
    ctx.body = Result.success(null, "已移除成员");
  }
}

export default new OrganizationController();

File: java-server/src/main/java/com/todothink/modules/system/organization/controller/OrganizationController.java
package com.todothink.modules.system.organization.controller;

import com.todothink.core.result.Result;
import com.todothink.core.security.LoginUser;
import com.todothink.core.security.SecurityUtils;
import com.todothink.core.util.RoleUtil;
import com.todothink.modules.system.organization.service.OrganizationService;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/departments")
public class OrganizationController {

    private final OrganizationService organizationService;

    public OrganizationController(OrganizationService organizationService) {
        this.organizationService = organizationService;
    }

    @GetMapping("/tree")
    public Result<?> getTree() {
        return Result.success(organizationService.getTree());
    }

    @GetMapping("/users")
    public Result<?> getAllUsers() {
        return Result.success(organizationService.getAllUsers());
    }

    @GetMapping("/{id}/managers")
    public Result<?> getManagers(@PathVariable Integer id) {
        return Result.success(organizationService.getManagers(id));
    }

    @PostMapping("/{id}/managers")
    public Result<?> addManagers(@PathVariable Integer id, @RequestBody Map<String, Object> body) {
        if (!isSuperAdmin()) {
            return Result.error("无权限：仅超级管理员可添加管理者", 403);
        }
        List<Integer> userIds = castIntList(body.get("userIds"));
        if (userIds == null || userIds.isEmpty()) {
            return Result.error("参数错误");
        }
        organizationService.addManagers(id, userIds);
        return Result.success(null, "添加管理者成功");
    }

    @DeleteMapping("/{id}/managers/{userId}")
    public Result<?> removeManager(@PathVariable Integer id, @PathVariable Integer userId) {
        if (!isSuperAdmin()) {
            return Result.error("无权限：仅超级管理员可移除管理者", 403);
        }
        organizationService.removeManager(id, userId);
        return Result.success(null, "已移除管理者");
    }

    @GetMapping("/{id}/members")
    public Result<?> getMembers(@PathVariable Integer id) {
        return Result.success(organizationService.getMembers(id));
    }

    @PostMapping("/{id}/members")
    public Result<?> addMembers(@PathVariable Integer id, @RequestBody Map<String, Object> body) {
        Integer uid = SecurityUtils.currentUserId();
        if (!isSuperAdmin() && !organizationService.isManager(id, uid)) {
            return Result.error("无权限：您不是该部门的管理者", 403);
        }
        List<Integer> userIds = castIntList(body.get("userIds"));
        if (userIds == null || userIds.isEmpty()) {
            return Result.error("参数错误");
        }
        organizationService.addMembers(id, userIds);
        return Result.success(null, "添加成员成功");
    }

    @DeleteMapping("/{id}/members/{userId}")
    public Result<?> removeMember(@PathVariable Integer id, @PathVariable Integer userId) {
        Integer uid = SecurityUtils.currentUserId();
        if (!isSuperAdmin() && !organizationService.isManager(id, uid)) {
            return Result.error("无权限：您不是该部门的管理者", 403);
        }
        organizationService.removeMember(id, userId);
        return Result.success(null, "已移除成员");
    }

    private boolean isSuperAdmin() {
        LoginUser user = SecurityUtils.currentUser();
        return user != null && RoleUtil.isSuperAdminRoles(user.getRoles());
    }

    @SuppressWarnings("unchecked")
    private List<Integer> castIntList(Object val) {
        if (val instanceof List) {
            return (List<Integer>) val;
        }
        return null;
    }
}
